### ISSUE TITLE
support regex matching for `src_query_args` in vmauth

### ISSUE THREAD

**Issue by NishantSarraff** [VictoriaMetrics/VictoriaMetrics#6070]
### Is your feature request related to a problem? Please describe

My use case is that I have two separate VM-clusters: preprod, and prod. While scraping the metrics, I am adding an external label (monitor="preprod" for preprod and monitor="prod" for production).

The reason for having two VM clusters was to ensure that changes or workload in preprod do not affect the production component in nay manner. 

I want that when the query comes from Grafana and in the query, if there is this value monitor="prod", then it routes the request to production-vmselects. If the value is monitor="preprod", then it routes the request to preprod-vmselects.

### Describe the solution you'd like

I would like the regex match here instead of exact match:
https://github.com/VictoriaMetrics/VictoriaMetrics/blob/v1.100.0/app/vmauth/target_url.go#L94

**Comment by NishantSarraff** [comment]
Example:

Request URL made by the grafana to VMAuth:
```
http://<base_url>/select/0/prometheus/api/v1/query_range?end=1234&query=http_request{monitor=preprod}&start=12213&step=30
```

VMAuth Version: v1.100.0

**Comment by hagen1778** [comment]
Hello @NishantSarraff!
The feature request sounds good, but one thing bothers me. Grafana, by default, will put `query` param into request body and will send it via POST method. Reading params from body isn't a good idea for proxies, since body can be read only once. If this request is meant to be proxied further, vmauth will have to re-create the body from scratch, so the destination backed could also read the body once received.
In two words, it is easy to add regex matching for GET params. But in most of cases, Grafana or vmui will put `query` param to body. And making vmauth to read the body and then recreate it could be pretty expensive for reverse-proxy.

**Comment by valyala** [comment]
There is yet another issue additionally to [the one described above](https://github.com/VictoriaMetrics/VictoriaMetrics/issues/6070#issuecomment-2056590909) with this feature request when trying to route incoming request based on the `query` passed to [/api/v1/query](https://docs.victoriametrics.com/keyconcepts/#instant-query) or [/api/v1/query_range](https://docs.victoriametrics.com/keyconcepts/#range-query) - this is very fragile, and it can miss proper routing of the incoming request for some edge cases. For example, if the following config is provided to `vmauth`:

```yaml
unauthorized_user:
  url_map:
  - src_query_args:
    - 'query=~.*monitor="prod".*'
    url_prefix: "http://prod-backend/"
```

Then requests with `query` containing `monitor="prod"` substring will be routed to `http://prod-backend/`. But this means that queries containing `<some_prefix>monitor="prod"` are also routed to `http://prod-backend/` (such as `foo_monitor="prod"`). This may be unexpected. It is also unclear where to route the request in the following cases:

- If the `query` doesn't match any provided regular expressions at `src_query_args`
- If the `query` matches multiple provided regular expressions at `src_query_args`
- If the `query` contains the needed substring, but this substring doesn't correspond to the actual [label filter](https://docs.victoriametrics.com/keyconcepts/#filtering). Instead, it may correspond to some label value. For example, `{foo='monitor="prod"'}`, e.g. this query returns time series with `foo` label containing `monitor="pod"` value.

Personally I don't like this particular use case for this feature request because it brings more potential issues comparing to the value it brings. There are other use cases when this feature can be used with less potential issues though...

P.S. VictoriaMetrics supports specifying additional [label filters](https://docs.victoriametrics.com/keyconcepts/#filtering) via `extra_labels` and `extra_filters` query args, which are automatically applied to all the queries it executes via all the supported querying APIs as described in [these docs](https://docs.victoriametrics.com/#prometheus-querying-api-enhancements). This allows routing queries in a more reliable way to different backends depending on the particular value for `extra_labels` /  `extra_filters` GET query arg.

**Comment by dennybaa** [comment]
Great thing to support regex in vmauth! Thank you @valyala. I wonder if the issue with the operator also could be improved, unfortunately VMUser configuration doesn't seem to have an ability to define the routing, such as the above `src_query_args` or multiple conditions?
